### PR TITLE
Data branch

### PR DESCRIPTION
**Pull request by melanyw**
Database updates.  The main interface class is CharacterRepository.

### CODE REVIEW

File: src/main/java/csci331/team/red/dao/CharacterDAO.java [hunk 1]
```diff
@@ -0,0 +1,462 @@
...
+	 * @param tableName
+	 * @return random id
+	 * @author melany
+	 */
+	public static int randomID(String tableName) {
+		PreparedStatement statement = null;
+		try {
+			statement = conn.connection.prepareStatement("SELECT ID FROM "
+					+ tableName + " ORDER BY Random() LIMIT 1");
+		} catch (SQLException e) {
```
Comment (ojourmel) [inline src/main/java/csci331/team/red/dao/CharacterDAO.java]: This exception is swallowed, but will continue to break the game. Perhaps just re-throw the exception as a `RuntimeException`.

If you think that an SQL error can be handled, or recovered from, then perhaps throw some kind of non-eql exception...

The same goes for all of the other SQLExceptions.

File: src/main/java/csci331/team/red/dao/CharacterDAO.java [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+	 * 
+	 * @param id
+	 * @return FirstName from database
+	 * @author melany
+	 */
+	public static String getFName(int id) {
+		String fName = "";
+
+		PreparedStatement statement = null;
+		try {
```
Comment (ojourmel) [inline src/main/java/csci331/team/red/dao/CharacterDAO.java]: These three `try` `catch` blocks could be amalgamated, if nothing else but to make the code nicer to read.

File: src/main/java/csci331/team/red/dao/CharacterRepository.java [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+	 * The construction creates an initial game-play character database
+	 * of 50 random characters
+	 * @author melany
+	 */
+	CharacterRepository() {
+		characters = new ArrayList<Character>();
+		CharacterDAO characterDAO = new CharacterDAO();
+
+		// Create the first 50 characters for game play
+		for (int i = 0; i < 50; i++) {
```
Comment (ojourmel) [inline src/main/java/csci331/team/red/dao/CharacterRepository.java]: Can `50` be replaced with a `private static final int INITAL_CHARACTERS=50`?

File: src/main/java/csci331/team/red/dao/CharacterRepository.java [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+							// matches
+		List<Character> chars = new LinkedList<Character>();
+
+		//Go through and see if the key is a database attribute
+		//If found, we add that character to the chars list
+		for (String s : map.keySet()) {
+			String val = map.get(s);
+
+			if (s == FIRSTNAME) {
+				for (int i = 0; i < characters.size(); i++) {
```
Comment (ojourmel) [inline src/main/java/csci331/team/red/dao/CharacterRepository.java]: This `for` loop can be replaced with an iterative for loop, just like is done with `map.keySet()`

I would recommend:

```
   for(Character character : characters){
      if(character.getFirstName() == val){
         chars.add(character);
      }
   }
```

File: src/main/java/csci331/team/red/dao/CharacterRepository.java [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+	 * <br>
+	 * If there are no characters in the game-play database that have
+	 * the required parameters, new characters will be created that have
+	 * the parameter, added to the character list AND to the game play database
+	 * and returned.
+	 * @param map
+	 * @return List of Characters, returns null on error
+	 */
+	public List<Character> getCharacters(Map<String, String> map) {
+		int newChars = 10; // number of new characters added to characters if no
```
Comment (ojourmel) [inline src/main/java/csci331/team/red/dao/CharacterRepository.java]: replace with a `private static final int`

File: src/main/java/csci331/team/red/dao/CharacterRepository.java [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+	 * there data is made to match the supplied attributes.  These new characters
+	 * are added both the the list of characters that are returned, and to the
+	 * game-play database of characters.
+	 * 
+	 * @param maps
+	 * @return List(Character)
+	 * 
+	 * @author melany
+	 */
+	// List<Map<String,Integer>> maps = new ArrayList<Map<String,Integer>>();
```
Comment (ojourmel) [inline src/main/java/csci331/team/red/dao/CharacterRepository.java]: Perhaps this was meant to be deleted?

File: src/main/java/csci331/team/red/dao/CharacterRepository.java [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+	public List<Character> getCharacters(List<Map<String, String>> maps) {
+		int newChars = 10;
```
Comment (ojourmel) [inline src/main/java/csci331/team/red/dao/CharacterRepository.java]: See my comment above about using a constant.

File: src/main/java/csci331/team/red/dao/CharacterRepository.java [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+		List<Character> chars = new LinkedList<Character>();
+
+		//Compare lists of characters with other lists of characters
+		//looking for duplicates for matches
+		for (int i = 0; i < (maps.size()); i++) {
+			List<Character> temp1 = getCharacters((maps.get(i++)));
+			List<Character> temp2 = getCharacters((maps.get(i)));
```
Comment (ojourmel) [inline src/main/java/csci331/team/red/dao/CharacterRepository.java]: Does this not have the possibility of going out of bounds? If `i == maps.size()-1`, then `i++` puts `i` out of array bounds. Is that possible?

**Comment by ojourmel** [comment]
~~Remove the `game.db` file that was added in the root of the repository.~~

Done

**Comment by melanyw** [comment]
re-factoring needed, will open again when done

**Comment by melanyw** [comment]
Value checking of database attributes, i.e. `dob` and `driversID`.
Addition of `Gender` and `PersonPicture` enum.
`getCharacter()` fixed to support multi-attribute searching

**Pull request merged**